### ISSUE TITLE
Call to decorator loses `this` context

### ISSUE THREAD

**Issue by monsanto** [babel/babel#1296]
Example:

```js
let x = {
    @foo.bar
    method() { }
}
```

I would expect the decorator call to desugar to `foo.bar(x, "method", ...)`, which ensures that the context of `bar` is `foo`. Instead, there is indirection between the property access and the function call, which loses the context:

```js
var x = _createDecoratedObject([{
    key: "method",
    decorators: [foo.bar],
    value: function value() {}
}]);
```

I looked at the [spec](https://github.com/wycats/javascript-decorators) and it isn't 100% clear to me what the right thing to do here is, as it only formally defines a grammar. However from [the examples](https://github.com/wycats/javascript-decorators#object-literal-method-declaration) I get the feeling that the intention is that the context of the call should be preserved, since the desugaring inlines the decorator for `F("color")` instead of assigning to an intermediate. Also, it is more useful!

**Comment by sebmck** [comment]
Yeah, a `bind` would be necessary here.

**Comment by sebmck** [comment]
Actually, I'm not 100% sure. cc @wycats

**Comment by wycats** [comment]
This is a pretty interesting case. I think the right thing is for it to desugar to a method call here. I plan to update the spec with a few cases people have raised next week, as well as to take a first stab at a formal grammar.

**Comment by sebmck** [comment]
Fixed as of 5.1.11, thanks!
